[docs] added soulbound NFT example (#22041)
MystenLabs/sui: docs/content/guides/developer/nft/nft-soulbound.mdx

---
title: Soulbound NFT Example
description: An example using Sui Move struct abilities and the Sui Framework's `transfer` module to make a NFT soulbound (non-transferable).
keywords: [ERC-721, NFT, Soulbound]
---

> A soulbound NFT is an NFT that is non-transferable. Once a NFT is minted to a Sui account, the NFT is "bounded" to that account and cannot be transferred. This implementation leverages the custom logic of the Sui framework's transfer functions. The [`sui::transfer` module](/references/framework/sui-framework/transfer.md) contains two functions that transfers objects: [`transfer::transfer`](/references/framework/sui-framework/transfer.md#function-transfer) and [`transfer::public_transfer`](/references/framework/sui-framework/transfer.md#function-public_transfer). 

Reviewer: ```suggestion
A soulbound NFT is an NFT that is non-transferable. After an NFT is minted to a Sui account, the NFT is "bounded" to that account and cannot be transferred. This implementation leverages the custom logic of the Sui framework's transfer functions. The <a href="/references/framework/sui-framework/transfer" data-noBrokenLinkCheck='true'>`sui::transfer` module</a> contains two functions that transfers objects: <a href="/references/framework/sui-framework/transfer#function-transfer" data-noBrokenLinkCheck='true'>`transfer::transfer`</a> and <a href="/references/framework/sui-framework/transfer#function-public_transfer" data-noBrokenLinkCheck='true'>`transfer::public_transfer`</a>. 
```
As you mentioned, the links break the build because the source pages don't exist at the time of checking.

Typically, when defining new NFTs or object types on Sui, you don't need to create a transfer function because the Sui Framework offers `transfer::public_transfer` which anyone can use to transfer objects. However, `transfer::public_transfer` requires transferred objects have the `key` and `store` ability. Therefore, if you define a new NFT type that has the `key` ability, meaning it is a Sui object, but not the `store` ability, the holders won't be able to use `transfer::public_transfer`. This results in a soulbound NFT.

> It is also possible to create custom transfer logic for NFTs on Sui. The `transfer::transfer` function has custom rules performed by the Sui Move bytecode verifier that ensures that the transferred objects is an object defined in the module where transfer is invoked. While removing the `store` ability from a struct definition makes `transfer::public_transfer` unusable, `transfer::transfer` can still be used as long as it's used in the module that defined that object's type. This allows for the module owner to provide custom transfer logic for their soulbound NFTs. 

Reviewer: ```suggestion
It is also possible to create custom transfer logic for NFTs on Sui. The `transfer::transfer` function has custom rules performed by the Sui Move bytecode verifier that ensures that the transferred objects are defined in the module where transfer is invoked. While removing the `store` ability from a struct definition makes `transfer::public_transfer` unusable, `transfer::transfer` can still be used as long as it's used in the module that defined that object's type. This allows for the module owner to provide custom transfer logic for their soulbound NFTs. 
```
Point of this suggestion is to make the plural matchy matchy. You could also choose to make "the transferred objects" singular instead.

## Example

The following example creates a basic soulbound NFT on Sui. The `TestnetSoulboundNFT` struct defines the NFT with an `id`, `name`, `description`, and `url` fields. 

{@inject: examples/move/nft-soulbound/sources/testnet_soulbound_nft.move#struct=TestnetSoulboundNFT noComments}

> This `TestnetSoulboundNFT` struct is defined with the `key` ability but without the `store` ability meaning it cannot be transferred with `transfer::public_transfer` and can only be transferred with `transfer::transfer`, which requires custom transfer logic to be implemented in the same module. 

Reviewer: ```suggestion
This `TestnetSoulboundNFT` struct is defined with the `key` ability but without the `store` ability. This means you cannot transfer it with `transfer::public_transfer`. Instead, use `transfer::transfer` with custom transfer logic implemented in the same module. 
```
Breaking up the long sentence. Make sure I didn't change the meaning.

> This example also includes an example of how to provide custom transfer logic using the `transfer::transfer` function. This is where you can add additional logic such as reset the NFT's stats, or requiring a payment. Don't provide this functionality if the NFT is fully soulbound. 

Reviewer: ```suggestion
This example also shows how to provide custom transfer logic using the `transfer::transfer` function. This is where you can add additional logic, such as resetting the NFT's stats or requiring a payment. Don't provide this functionality if the NFT is fully soulbound. 
```

{@inject: examples/move/nft-soulbound/sources/testnet_soulbound_nft.move#fun=transfer}

<details>
<summary>
`testnet_soulbound_nft.move` 
</summary>

{@inject: examples/move/nft-soulbound/sources/testnet_soulbound_nft.move}

</details>

## Related links

> - [Soulbound NFT Example source code](https://github.com/MystenLabs/sui/tree/main/examples/move/nft-soulbound): The source code that this document references.

Reviewer: ```suggestion
- [Soulbound NFT example source code](https://github.com/MystenLabs/sui/tree/main/examples/move/nft-soulbound): The source code that this document references.
```
Nittiest of nits. I see why you formatted it the way you did, but I think this is ever so slightly more correct.
